<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.433">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Krysten Jones (A10553682)">

<title>Find_Gene_Project</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="Find_Gene_Project_files/libs/clipboard/clipboard.min.js"></script>
<script src="Find_Gene_Project_files/libs/quarto-html/quarto.js"></script>
<script src="Find_Gene_Project_files/libs/quarto-html/popper.min.js"></script>
<script src="Find_Gene_Project_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="Find_Gene_Project_files/libs/quarto-html/anchor.min.js"></script>
<link href="Find_Gene_Project_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="Find_Gene_Project_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="Find_Gene_Project_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="Find_Gene_Project_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="Find_Gene_Project_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Find_Gene_Project</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Krysten Jones (A10553682) </p>
          </div>
  </div>
    
  
    
  </div>
  

</header>

<section id="using-bio3d" class="level1">
<h1>Using Bio3D</h1>
<blockquote class="blockquote">
<p>[Q7] Generate a sequence identity based heatmap of your aligned sequences using R.If necessary convert your sequence alignment to the ubiquitous FASTA format (Seaview can read in clustal format and “Save as” FASTA format for example). Read this FASTA format alignment into R with the help of functions in the Bio3D package.</p>
</blockquote>
<p>First as always, call your programs</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(bio3d)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>Fasta <span class="ot">&lt;-</span> <span class="fu">read.fasta</span>(<span class="st">"MEGA_FASTA.fas"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<blockquote class="blockquote">
<p>Calculate a sequence identity matrix (again using a function within the Bio3D package).</p>
</blockquote>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>identity_matrix <span class="ot">&lt;-</span> <span class="fu">seqidentity</span>(Fasta)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>identity_matrix</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                            T._guttata_sialytransferase Human_st6gal1
T._guttata_sialytransferase                       1.000         0.014
Human_st6gal1                                     0.014         1.000
Green_crombec                                     0.965         0.014
California_thrasher                               0.950         0.017
Black_sunbird                                     0.965         0.014
Elegant_woodcreeper                               0.930         0.014
Loggerhead_shrike                                 0.953         0.014
Softshell_turtle                                  0.837         0.043
American_alligator                                0.802         0.043
Loggerhead_sea_turtle                             0.814         0.043
Vesper_bat                                        0.791         0.014
Indian_flying_fox                                 0.814         0.014
                            Green_crombec California_thrasher Black_sunbird
T._guttata_sialytransferase         0.965               0.950         0.965
Human_st6gal1                       0.014               0.017         0.014
Green_crombec                       1.000               0.983         0.977
California_thrasher                 0.983               1.000         1.000
Black_sunbird                       0.977               1.000         1.000
Elegant_woodcreeper                 0.942               0.983         0.965
Loggerhead_shrike                   0.965               0.967         0.965
Softshell_turtle                    0.872               0.867         0.860
American_alligator                  0.837               0.900         0.837
Loggerhead_sea_turtle               0.849               0.900         0.849
Vesper_bat                          0.814               0.850         0.802
Indian_flying_fox                   0.837               0.850         0.826
                            Elegant_woodcreeper Loggerhead_shrike
T._guttata_sialytransferase               0.930             0.953
Human_st6gal1                             0.014             0.014
Green_crombec                             0.942             0.965
California_thrasher                       0.983             0.967
Black_sunbird                             0.965             0.965
Elegant_woodcreeper                       1.000             0.977
Loggerhead_shrike                         0.977             1.000
Softshell_turtle                          0.860             0.872
American_alligator                        0.860             0.849
Loggerhead_sea_turtle                     0.860             0.849
Vesper_bat                                0.837             0.837
Indian_flying_fox                         0.860             0.860
                            Softshell_turtle American_alligator
T._guttata_sialytransferase            0.837              0.802
Human_st6gal1                          0.043              0.043
Green_crombec                          0.872              0.837
California_thrasher                    0.867              0.900
Black_sunbird                          0.860              0.837
Elegant_woodcreeper                    0.860              0.860
Loggerhead_shrike                      0.872              0.849
Softshell_turtle                       1.000              0.814
American_alligator                     0.814              1.000
Loggerhead_sea_turtle                  0.849              0.802
Vesper_bat                             0.767              0.767
Indian_flying_fox                      0.791              0.767
                            Loggerhead_sea_turtle Vesper_bat Indian_flying_fox
T._guttata_sialytransferase                 0.814      0.791             0.814
Human_st6gal1                               0.043      0.014             0.014
Green_crombec                               0.849      0.814             0.837
California_thrasher                         0.900      0.850             0.850
Black_sunbird                               0.849      0.802             0.826
Elegant_woodcreeper                         0.860      0.837             0.860
Loggerhead_shrike                           0.849      0.837             0.860
Softshell_turtle                            0.849      0.767             0.791
American_alligator                          0.802      0.767             0.767
Loggerhead_sea_turtle                       1.000      0.767             0.767
Vesper_bat                                  0.767      1.000             0.930
Indian_flying_fox                           0.767      0.930             1.000</code></pre>
</div>
</div>
<blockquote class="blockquote">
<p>Then generate a heatmap plot and add to your report. Do make sure your labels are visible and not cut at the ﬁgure margins.</p>
</blockquote>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"pheatmap"</span>)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="co">#vector of labels</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>seq_labs <span class="ot">&lt;-</span> <span class="fu">c</span>( <span class="st">"T._guttata_sialytransferase"</span>,<span class="st">"Human_St6Gal1"</span>, <span class="st">"Green_crombec"</span>, <span class="st">"California_thrasher"</span>, <span class="st">"Black_sunbird"</span>, <span class="st">"Elegant_woodcreeper"</span>, <span class="st">"Loggerhead_shrike"</span>, <span class="st">"Softshell_turtle"</span>, <span class="st">"American_alligator"</span>, <span class="st">"Loggerhead_sea_turtle"</span>, <span class="st">"Vesper_bat"</span>, <span class="st">"Indian_flying_fox"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Make heatmap using identity matrix</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">pheatmap</span>(identity_matrix,</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>         <span class="at">cluster_rows =</span> <span class="cn">TRUE</span>,</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>         <span class="at">cluster_cols =</span> <span class="cn">TRUE</span>,</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>         <span class="at">annotation_legend_title =</span> <span class="st">"Sequence Identity Heatmap"</span>,</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>         <span class="at">color=</span><span class="fu">colorRampPalette</span>(<span class="fu">c</span>(<span class="st">"red"</span>, <span class="st">"yellow"</span>))(<span class="dv">100</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="Find_Gene_Project_files/figure-html/unnamed-chunk-4-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<section id="generating-a-consensus-sequence-and-row-wise-maximum-from-your-sequence-identity-matrix" class="level2">
<h2 class="anchored" data-anchor-id="generating-a-consensus-sequence-and-row-wise-maximum-from-your-sequence-identity-matrix">Generating a consensus sequence and row-wise maximum from your sequence identity matrix</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co">#determining file type</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(Fasta)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "fasta"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">file.info</span>(<span class="st">"MEGA_FASTA.fas"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>               size isdir mode               mtime               ctime
MEGA_FASTA.fas 2281 FALSE  666 2023-12-07 12:23:31 2023-12-07 12:10:50
                             atime exe
MEGA_FASTA.fas 2023-12-07 20:01:17  no</code></pre>
</div>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># generating consensus sequence using Bio3D</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>con_seq <span class="ot">&lt;-</span> <span class="fu">consensus</span>(Fasta)</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(con_seq<span class="sc">$</span>seq)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "P" "S" "S" "G" "M" "L" "G" "I" "-" "-" "M" "M" "T" "L" "C" "D" "-" "V" "D"
[20] "V" "Y" "E" "F" "L" "P" "S" "K" "R" "Q" "T" "D" "I" "C" "H" "Y" "Y" "Q" "K"
[39] "F" "H" "D" "-" "A" "C" "T" "M" "G" "A" "Y" "H" "P" "L" "L" "F" "E" "K" "N"
[58] "L" "V" "K" "-" "-" "N" "Q" "G" "T" "D" "E" "-" "I" "Y" "-" "-" "G" "K" "V"
[77] "T" "L" "P" "G" "F" "R" "-" "-" "H" "C"</code></pre>
</div>
</div>
<p>What if we change the cutoff, does that change the sequence?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>con <span class="ot">&lt;-</span> <span class="fu">consensus</span>(Fasta, <span class="at">cutoff =</span> <span class="fl">0.6</span>)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(con<span class="sc">$</span>seq)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "P" "S" "S" "G" "M" "L" "G" "I" "-" "-" "M" "M" "T" "L" "C" "D" "-" "V" "D"
[20] "V" "Y" "E" "F" "L" "P" "S" "K" "R" "Q" "T" "D" "I" "C" "H" "Y" "Y" "Q" "K"
[39] "F" "H" "D" "-" "A" "C" "T" "M" "G" "A" "Y" "H" "P" "L" "L" "F" "E" "K" "N"
[58] "L" "V" "K" "-" "-" "N" "Q" "G" "T" "D" "E" "-" "I" "Y" "-" "-" "G" "K" "V"
[77] "T" "L" "P" "G" "F" "R" "-" "-" "H" "C"</code></pre>
</div>
</div>
<p>This appears to be the default cutoff</p>
</section>
<section id="blast-pdb" class="level2">
<h2 class="anchored" data-anchor-id="blast-pdb">Blast pdb</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>consensus_blast <span class="ot">&lt;-</span> <span class="fu">blast.pdb</span>(con<span class="sc">$</span>seq)</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>hits <span class="ot">&lt;-</span> <span class="fu">plot</span>(consensus_blast)</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Then save it so we don't have to blast again and you can skip to the next code chunk</span></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a><span class="fu">save</span>(hits, consensus_blast, <span class="at">file =</span> <span class="st">"consensus_blast_results.Rds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Using the <code>blast.pdb()</code> function with the consensus sequence threw an error the first time so I did a bunch of extra unnecessary steps below TT.TT. Skip from here to “PDB Results” section</p>
<p>Now we can load the results and skip re-running blast through the code chunk above</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">"consensus_blast_results.Rds"</span>)</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>hits</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$hits
  pdb.id   acc      group
1 "6QVS_A" "6QVS_A" "1"  
2 "4JS1_A" "4JS1_A" "1"  
3 "4MPS_A" "4MPS_A" "1"  

$pdb.id
[1] "6QVS_A" "4JS1_A" "4MPS_A"

$acc
[1] "6QVS_A" "4JS1_A" "4MPS_A"

$inds
[1]  TRUE  TRUE  TRUE FALSE FALSE FALSE FALSE FALSE

attr(,"class")
[1] "blast"</code></pre>
</div>
</div>
</section>
<section id="skip-if-consensus-search-works" class="level2">
<h2 class="anchored" data-anchor-id="skip-if-consensus-search-works">Skip if consensus search works</h2>
<p>So it would be better just to search the sequence of interest and the highest conserved sequence based on the sequence identity matrix.</p>
<p>To do this, we first need to subset just our sequence of interest from the FASTA file.</p>
</section>
<section id="identity-matrix-max" class="level2">
<h2 class="anchored" data-anchor-id="identity-matrix-max">identity matrix max</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>row_max_values <span class="ot">&lt;-</span> <span class="fu">apply</span>(identity_matrix, <span class="dv">1</span>, max)</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>row_max_values</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Also having all the max be 1 is odd, this is because I’ve set the same identity to itself for the heatmap, we’re interested in the other identities for this, lets look at all of the unique identities</p>
<p>Finding identity matrix unique</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>unique_values_per_row <span class="ot">&lt;-</span> <span class="fu">apply</span>(identity_matrix, <span class="dv">1</span>, unique)</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>unique_values_per_row</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Trying to get an easier visualization so I don’t have to look at all the numbers</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>greater_than_threshold <span class="ot">&lt;-</span> identity_matrix <span class="sc">&gt;=</span> <span class="fl">0.983</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>greater_than_threshold</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>values_above_threshold <span class="ot">&lt;-</span> identity_matrix[greater_than_threshold]</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(values_above_threshold)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Green crombec, california thrasher, elegant woodcreeper, and black sunbird all have the highest sequence identity of 0.983. So any of these sequences should work</p>
</section>
<section id="subsetting-a-fasta-file-double-skip-this" class="level2">
<h2 class="anchored" data-anchor-id="subsetting-a-fasta-file-double-skip-this">Subsetting a FASTA file (double skip this)</h2>
<p>This is only if you forget quotation marks for your FASTA sequence so you think something is wrong &gt;.&lt;</p>
<p>To subset a fasta file, you first need to define the file path. You can figure out and set the file path using <code>getwd()</code> and assigning it to a variable.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">getwd</span>()</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>file_path <span class="ot">&lt;-</span> <span class="st">"/Users/kryst/Documents/UCSD 2022/Classes/2023 Q4 Fall/BGGN 213/Lab Assignments/Find_Gene_Project/MEGA_FASTA.fas"</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Subsetting sequence from fasta file using Biostrings package. Biostrings is part of bioconductor so first you need to install it using <code>BiocManager::install("Biostrings")</code>. Looking at our original FASTA file, the sequence of interest is stored as the third sequence so in the <code>fasta_sequences</code> function, we will call the third vector</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Biostrings)</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>fasta_sequences <span class="ot">&lt;-</span><span class="fu">readAAStringSet</span>(file_path)</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>guttata <span class="ot">&lt;-</span> fasta_sequences[[<span class="dv">3</span>]]</span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>guttata</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="getting-protein-database-pdb-info-using-bio3d-skip-again-if-consensus-works" class="level2">
<h2 class="anchored" data-anchor-id="getting-protein-database-pdb-info-using-bio3d-skip-again-if-consensus-works">Getting Protein Database (PDB) info using Bio3D (skip again if consensus works)</h2>
<p>The <code>blast.pdb()</code> function will only use a single sequence character, but our subsetting above didn’t work because there is other data. Instead you can just post the FASTA sequence using quotation marks</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>guttata <span class="ot">&lt;-</span> <span class="fu">blast.pdb</span>(<span class="st">"PSSGMLGIVLMXXLCDQVDVYEFLPSKRQTDICHYYQKFHDHACTMGAYHPLLFEKNLVKHMNQGTDEDIYTHGKVTLPGFRKVHC"</span>)</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="co"># save output so we don't need to run blast every time</span></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>g_hits <span class="ot">&lt;-</span> <span class="fu">plot</span>(guttata)</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a><span class="fu">save</span>(g_hits, guttata, <span class="at">file =</span> <span class="st">"guttata_blast_results.Rds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Doing the same using black sunbird as the consensus due to it having the highest sequence max identity (in this case 0.983). Could have picked one of the 3 others with this value as well</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>bird <span class="ot">&lt;-</span> <span class="fu">blast.pdb</span>(<span class="st">"PSSGMLGIVLMMTLCDEVDVYEFLPSKRQTDICHYYQKFHDHACTMGAYHPLLFEKNLVKHMNQGTDEDIYTHGKVTLPGFRKVHC"</span>)</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>b_hits <span class="ot">&lt;-</span> <span class="fu">plot</span>(bird)</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a><span class="fu">save</span>(b_hits, bird, <span class="at">file =</span> <span class="st">"bird_blast_results.Rds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Interestingly, when you compare the T.guttata sequence to inputting the FASTA sequence in blastp (https://blast.ncbi.nlm.nih.gov/) and using the pdb database, you get 7 hits instead of 3.</p>
<p>Loading pdb files we made</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">"guttata_blast_results.Rds"</span>)</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>g_hits</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">"bird_blast_results.Rds"</span>)</span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>b_hits</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="pdb-results" class="level2">
<h2 class="anchored" data-anchor-id="pdb-results">PDB Results</h2>
<p>The results from CONSENSUS online search include:</p>
<p>Getting results from CONSENSUS Bio3D search include:6QVS, 4JS1, and 4MPS which is the same as just searching the guttata sequence (see further code chunks)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co"># pull files</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>files <span class="ot">&lt;-</span> <span class="fu">get.pdb</span>(hits<span class="sc">$</span>pdb.id, <span class="at">path=</span><span class="st">"pdbs"</span>, <span class="at">split=</span><span class="cn">TRUE</span>, <span class="at">gzip=</span><span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in get.pdb(hits$pdb.id, path = "pdbs", split = TRUE, gzip = TRUE):
pdbs/6QVS.pdb exists. Skipping download</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in get.pdb(hits$pdb.id, path = "pdbs", split = TRUE, gzip = TRUE):
pdbs/4JS1.pdb exists. Skipping download</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in get.pdb(hits$pdb.id, path = "pdbs", split = TRUE, gzip = TRUE):
pdbs/4MPS.pdb exists. Skipping download</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>
  |                                                                            
  |                                                                      |   0%
  |                                                                            
  |=======================                                               |  33%
  |                                                                            
  |===============================================                       |  67%
  |                                                                            
  |======================================================================| 100%</code></pre>
</div>
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="co"># align files</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>pdbs <span class="ot">&lt;-</span> <span class="fu">pdbaln</span>(files, <span class="at">fit =</span> <span class="cn">TRUE</span>, <span class="at">exefile=</span><span class="st">"msa"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Reading PDB files:
pdbs/split_chain/6QVS_A.pdb
pdbs/split_chain/4JS1_A.pdb
pdbs/split_chain/4MPS_A.pdb
   PDB has ALT records, taking A only, rm.alt=TRUE
...

Extracting sequences

pdb/seq: 1   name: pdbs/split_chain/6QVS_A.pdb 
   PDB has ALT records, taking A only, rm.alt=TRUE
pdb/seq: 2   name: pdbs/split_chain/4JS1_A.pdb 
pdb/seq: 3   name: pdbs/split_chain/4MPS_A.pdb </code></pre>
</div>
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="co"># create id's for figure</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>ids <span class="ot">&lt;-</span> <span class="fu">basename.pdb</span>(pdbs<span class="sc">$</span>id)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You can use this to plot a giant figure</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(pdbs, <span class="at">labels=</span>ids)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The results from our guttata online search include: 4JS1, 6QVS, and 4MPS</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>js1 <span class="ot">&lt;-</span> <span class="fu">read.pdb</span>(<span class="st">"4JS1"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Note: Accessing on-line PDB file</code></pre>
</div>
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>qvs <span class="ot">&lt;-</span> <span class="fu">read.pdb</span>(<span class="st">"6QVS"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Note: Accessing on-line PDB file
   PDB has ALT records, taking A only, rm.alt=TRUE</code></pre>
</div>
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>mps <span class="ot">&lt;-</span> <span class="fu">read.pdb</span>(<span class="st">"4MPS"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Note: Accessing on-line PDB file</code></pre>
</div>
</div>
<p>The results from the Bio3D search for T. guttata are: 4JS1, 6QVS, 4MPS, 6APL, 6APJ, 2WML, and 2WNB</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="co"># same 3 as above, but also</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>apl <span class="ot">&lt;-</span><span class="fu">read.pdb</span>(<span class="st">"5APL"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Note: Accessing on-line PDB file
   PDB has ALT records, taking A only, rm.alt=TRUE</code></pre>
</div>
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>apj <span class="ot">&lt;-</span><span class="fu">read.pdb</span>(<span class="st">"6APJ"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Note: Accessing on-line PDB file</code></pre>
</div>
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>wml <span class="ot">&lt;-</span><span class="fu">read.pdb</span>(<span class="st">"2WML"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Note: Accessing on-line PDB file
   PDB has ALT records, taking A only, rm.alt=TRUE</code></pre>
</div>
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>wnb <span class="ot">&lt;-</span><span class="fu">read.pdb</span>(<span class="st">"2WNB"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Note: Accessing on-line PDB file
   PDB has ALT records, taking A only, rm.alt=TRUE</code></pre>
</div>
</div>
<p>The results from our online search using “bird” include: 4JS1, 6QVS, 4MPS, 2WML, and 2WNB which are included in the Bio3d search for T. guttata sequence</p>
<p>The results from the Bio3D search for “bird” are:</p>
<p>Let’s take a look at some of the information stored in here</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>js1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
 Call:  read.pdb(file = "4JS1")

   Total Models#: 1
     Total Atoms#: 2768,  XYZs#: 8304  Chains#: 2  (values: A B)

     Protein Atoms#: 2594  (residues/Calpha atoms#: 318)
     Nucleic acid Atoms#: 0  (residues/phosphate atoms#: 0)

     Non-protein/nucleic Atoms#: 174  (residues: 52)
     Non-protein/nucleic resid values: [ BMA (1), CTN (1), GAL (2), HOH (41), MAN (2), NAG (4), PO4 (1) ]

   Protein sequence:
      PEASFQVWNKDSSSKNLIPRLQKIWKNYLSMNKYKVSYKGPGPGIKFSAEALRCHLRDHV
      NVSMVEVTDFPFNTSEWEGYLPKESIRTKAGPWGRCAVVSSAGSLKSSQLGREIDDHDAV
      LRFNGAPTANFQQDVGTKTTIRLMNSQLVTTEKRFLKDSLYNEGILIVWDPSVYHSDIPK
      WYQNPDYNFFNNYKTYRKLHPNQPFYILKPQMPWELWDILQEISP...&lt;cut&gt;...TIHC

+ attr: atom, xyz, seqres, helix, sheet,
        calpha, remark, call</code></pre>
</div>
</div>
<p>Cool, lets grab some more data so we can do some conversions</p>
<p>First download related pdb files, not sure how this is different from running the blast search above, I think it’s just calling part of it? specifically it’s calling the pdb id’s from the hits files which we can do instead of calling each of the ID’s individually like I do above</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="co"># pull files</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>g_files <span class="ot">&lt;-</span> <span class="fu">get.pdb</span>(g_hits<span class="sc">$</span>pdb.id, <span class="at">path=</span><span class="st">"pdbs"</span>, <span class="at">split=</span><span class="cn">TRUE</span>, <span class="at">gzip=</span><span class="cn">TRUE</span>)</span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a>b_files <span class="ot">&lt;-</span> <span class="fu">get.pdb</span>(b_hits<span class="sc">$</span>pdb.id, <span class="at">path=</span><span class="st">"pdbs"</span>, <span class="at">split=</span><span class="cn">TRUE</span>, <span class="at">gzip=</span><span class="cn">TRUE</span>)</span>
<span id="cb50-4"><a href="#cb50-4" aria-hidden="true" tabindex="-1"></a><span class="co"># align files</span></span>
<span id="cb50-5"><a href="#cb50-5" aria-hidden="true" tabindex="-1"></a>pdbs <span class="ot">&lt;-</span> <span class="fu">pdbaln</span>(files, <span class="at">fit =</span> <span class="cn">TRUE</span>, <span class="at">exefile=</span><span class="st">"msa"</span>)</span>
<span id="cb50-6"><a href="#cb50-6" aria-hidden="true" tabindex="-1"></a><span class="co"># create id's for figure</span></span>
<span id="cb50-7"><a href="#cb50-7" aria-hidden="true" tabindex="-1"></a>ids <span class="ot">&lt;-</span> <span class="fu">basename.pdb</span>(pdbs<span class="sc">$</span>id)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="calculating-rmsd" class="level2">
<h2 class="anchored" data-anchor-id="calculating-rmsd">Calculating RMSD</h2>
<p><code>pdb.annotate()</code> provides a convenient way of annotating the PDB files we have collected. Below we use the function to annotate each structure to its source species. This will come in handy when annotating plots later on</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate RMSD</span></span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>rd <span class="ot">&lt;-</span> <span class="fu">rmsd</span>(pdbs)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>